store-implementation/accumulo-store/src/main/java/uk/gov/gchq/gaffer/accumulostore/key/core/impl/byteEntity/ByteEntityAccumuloElementConverter.java

```diff
@@ -61,57 +60,52 @@ public ByteEntityAccumuloElementConverter(final Schema schema) {
-            directionFlag2 = ByteEntityPositions.INCORRECT_WAY_DIRECTED_EDGE;
-        } else {
-            directionFlag1 = ByteEntityPositions.UNDIRECTED_EDGE;
-            directionFlag2 = ByteEntityPositions.UNDIRECTED_EDGE;
-        }
-        final byte[] source = getSerialisedSource(edge);
-        final byte[] destination = getSerialisedDestination(edge);
-
-        final int length = source.length + destination.length + 5;
-        final byte[] rowKey1 = new byte[length];
-        System.arraycopy(source, 0, rowKey1, 0, source.length);
-        rowKey1[source.length] = ByteArrayEscapeUtils.DELIMITER;
-        rowKey1[source.length + 1] = directionFlag1;
-        rowKey1[source.length + 2] = ByteArrayEscapeUtils.DELIMITER;
-        System.arraycopy(destination, 0, rowKey1, source.length + 3, destination.length);
-        rowKey1[rowKey1.length - 2] = ByteArrayEscapeUtils.DELIMITER;
-        rowKey1[rowKey1.length - 1] = directionFlag1;
-        final byte[] rowKey2 = new byte[length];
-        System.arraycopy(destination, 0, rowKey2, 0, destination.length);
-        rowKey2[destination.length] = ByteArrayEscapeUtils.DELIMITER;
-        rowKey2[destination.length + 1] = directionFlag2;
-        rowKey2[destination.length + 2] = ByteArrayEscapeUtils.DELIMITER;
-        System.arraycopy(source, 0, rowKey2, destination.length + 3, source.length);
-        rowKey2[rowKey2.length - 2] = ByteArrayEscapeUtils.DELIMITER;
-        rowKey2[rowKey2.length - 1] = directionFlag2;
-        if (selfEdge(edge)) {
-            return new Pair<>(rowKey1, null);
+        byte[] source = getSerialisedSource(edge);
+        byte[] destination = getSerialisedDestination(edge);
+
+        byte directionFlag = edge.isDirected() ? ByteEntityPositions.CORRECT_WAY_DIRECTED_EDGE : ByteEntityPositions.UNDIRECTED_EDGE;
+        final byte[] rowKey1 = getRowKey(source, destination, directionFlag);
+
+        byte[] rowKey2 = null;
+        if (!selfEdge(edge)) {
+            byte invertDirectedFlag = (directionFlag == ByteEntityPositions.CORRECT_WAY_DIRECTED_EDGE) ? ByteEntityPositions.INCORRECT_WAY_DIRECTED_EDGE : directionFlag;
+            rowKey2 = getRowKey(destination, source, invertDirectedFlag);
         }
+
         return new Pair<>(rowKey1, rowKey2);
     }
 
+    private byte[] getRowKey(final byte[] first, final byte[] second, final byte directionFlag) {
+        int carriage = first.length;
+        int secondLen = second.length;
+        byte[] rowKey = new byte[carriage + secondLen + 5];
+        System.arraycopy(first, 0, rowKey, 0, carriage);
+        rowKey[carriage++] = ByteArrayEscapeUtils.DELIMITER;
+        rowKey[carriage++] = directionFlag;
+        rowKey[carriage++] = ByteArrayEscapeUtils.DELIMITER;
+        System.arraycopy(second, 0, rowKey, carriage, secondLen);
+        carriage += secondLen;
+        rowKey[carriage++] = ByteArrayEscapeUtils.DELIMITER;
+        rowKey[carriage] = directionFlag;
+        //carriage++;
```

Comment:
I had kept this comment in there as a visual clue that if the carriage is to be used again it needs appending.